README.md
share2nightscout-bridge
=======================

[![wercker status](https://app.wercker.com/status/1d9a86d110cb9d42c844fa60d084e5c4/m "wercker status")](https://app.wercker.com/project/bykey/1d9a86d110cb9d42c844fa60d084e5c4)
[![Gitter chat][gitter-img]][gitter-url]
[Releases][releases]

[![Deploy to Heroku][heroku-img]][heroku-url]

[releases]: https://github.com/bewest/share2nightscout-bridge/releases
[heroku-img]: https://www.herokucdn.com/deploy/button.png
[heroku-url]: https://heroku.com/deploy
[gitter-img]: https://img.shields.io/badge/Gitter-Join%20Chat%20%E2%86%92-1dce73.svg
[gitter-url]: https://gitter.im/nightscout/public
[c-r-m]: https://github.com/nightscout/cgm-remote-monitor
[wiki]: https://github.com/bewest/share2nightscout-bridge/wiki
[dexcom-eula]: http://www.dexcom.com/node/5421
[azure-environment]: https://github.com/projectkudu/kudu/wiki/Azure-runtime-environment
[blog-post]: http://www.hanselman.com/blog/BridgingDexcomShareCGMReceiversAndNightscout.aspx

The Share to Nightscout bridge copies your CGM data from Dexcom web services to
a Nightscout website.  The bridge runs as `node.js index.js` and will loop
forever, periodically querying Dexcom's Share web services for new CGM data.
The bridge relays any new data to a Nightscout website
([cgm-remote-monitor][c-r-m]) via the REST API.  The website then stores the
data in a Mongo database.

### Prerequisites

* A working Dexcom Share receiver paired to an Apple device that is
  successfully uploading data to Dexcom.  You must be able to see the Dexcom
  data in the Dexcom Follow app for the bridge to work.
* Your Dexcom Sharer username and password
* A working Nightscout website and Mongo database

### Install

The Share to Nightscout bridge is supported on both Azure and Heroku.  Please
see the [wiki][wiki] for current install information.

### Environment

`VARIABLE` (default) - description

#### Required

* `API_SECRET` - A secret passphrase that must be at least 12 characters long, and must match the `API_SECRET` from your Nightscout website
* `DEXCOM_ACCOUNT_NAME` - Your Dexcom Share2 username
* `DEXCOM_PASSWORD` - Your Dexcom Share2 password
* `WEBSITE_HOSTNAME` - The host name for your Nightscout website.  Example: sitename.herokuapp.com or sitename.azurewebsites.net.

#### Optional

* `maxCount` (1) - The maximum number of records to fetch per update
* `minutes` (1440) - The time window to search for new data per update (default is one day in minutes)
* `firstFetchCount` (3) - Changes `maxCount` during the very first update only.
* `maxFailures` (3) - The program will stop running after this many
  consecutively failed login attempts with a clear error message in the logs.
* `SHARE_INTERVAL` (150000) - The time to wait between each update (default is 2.5 minutes in milliseconds)
* `NS` - A fully-qualified Nightscout URL (e.g. `https://sitename.herokuapp.com`) which overrides `WEBSITE_HOSTNAME`

#### Azure Specific

* It is highly recommended that you set the `API_SECRET`, `DEXCOM_ACCOUNT_NAME` and `DEXCOM_PASSWORD` in **Connection Strings**.
* No need to set `WEBSITE_HOSTNAME` because the value is obtained from the existing [Azure website environment][azure-environment].

### More information

[As described by Scott Hanselman][blog-post], the bridge logs in to Dexcom
Share as the data publisher.  It re-uses the token every `5` minutes to fetch
the `maxCount` latest glucose records within the last specified `minutes`.
This information is then sent to the user's specified Nightscout install,
making the data available to the beloved pebble watch and other equipment owned
and operated by the receiver's owner.  It will continue to re-use the same
`sessionID` until it expires, at which point it should attempt to log in again.
If it can log in again, it will continue to re-use the new token to fetch data,
storing it into Nightscout.

This project is not FDA approved, not recommended for therapy, and not
recommended by [Dexcom][dexcom-eula].



History.md


0.2.1 / 2019-03-07
==================

  * Fix a bug where user can accidentally cause excessive API calls to be made, resulting in Nightscout consuming a huge amount of resources in the hosting environment (and the API calls probably failing)

0.2.0 / 2019-01-04
==================

  * Merge pull request #34 from nightscout/dev
  * release 0.2.0, also implement sulkaharo code review changes
  * release to npm as share2nightsout-bridge@0.2.0-dev-20190102 with tag dev
  * Merge branch 'master' into dev
  * integrate changes into dev branch for the 0.2.0 release
  * Merge pull request #33 from veryfancy/check-res
  * Check res before checking res.statusCode to prevent server crashes
  * Merge pull request #28 from PieterGit/wip/generalize
  * Update package.json
  * upgrade to 0.1.6
  * upgrade request and resolve security issues
  * Merge pull request #24 from mayaalinder/wip/share-servers
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js
  * Update index.js

v0.1.5 / 2015-09-26
===================

  * 0.1.5
  * Merge pull request #22 from mddub/wip/document-ns-env-var
  * Document "NS" environment variable
  * Merge pull request #20 from bewest/wip/generalize
  * only nullify_battery_status if endpoint is set
  * allow re-use via passing a callback

v0.1.4 / 2015-04-01
===================

  * 0.1.4
  * Merge pull request #14 from bewest/wip/reduce-bad-login-attempts
  * add maxFailures to README, tweak ws
  * handle login failures better

v0.1.3 / 2015-03-31
===================

  * 0.1.3
  * Merge pull request #12 from bewest/wip/battery-status
  * ensure first run is scoped soley to first run
  * use Defaults variable, eliminate globals
  * tweak startup behavior
  * Merge pull request #13 from jimsiff/wip/checks
  * moved checks above switch
  * introduce first run behavior
  * typo, missed )
  * Add crude email address check for DEXCOM_ACCOUNT_NAME
  * Add min API_SECRET length check
  * Merge pull request #3 from bewest/master
  * Merge pull request #8 from bewest/wip/add-heroku
  * Merge pull request #7 from jimsiff/wip/readme
  * intro clarifications
  * minor changes
  * A few fixes
  * Complete re-org
  * fixed link
  * New intro
  * Removed old button, etc
  * Move Deploy button
  * Merge pull request #2 from bewest/master
  * Merge pull request #6 from bewest/wip/add-heroku
  * Merge pull request #5 from jimsiff/wip/heroku-button
  * Optional vars not required to deploy
  * Updated name, optional vars, tweaked descriptions, removed Mongolab add-on, added Papertrail
  * clean up README, revert code changes
  * 1min
  * Update README.md
  * Update README.md
  * D2H_instructions
  * Update app.json
  * Update app.json
  * Update app.json
  * Update app.json
  * Update app.json
  * Update app.json
  * Update app.json
  * Update app.json
  * Update app.json
  * Update README.md
  * Update app.json
  * Update app.json
  * Update app.json
  * node
  * Update app.json
  * Update app.json
  * Update app.json
  * deploy_buttons
  * Create app.json
  * Create Procfile
  * clarify billing expecations

v0.1.2 / 2015-03-15
===================

  * 0.1.2
  * Change default to 2.5 minutes per nyquist

v0.1.1 / 2015-03-15
===================

  * 0.1.1
  * attempt to provide direction
  * tweak features
  * update docs

v0.1.0 / 2015-03-14
===================

  * 0.1.0
  * update to make easier to use
  * add badge

v0.0.3 / 2015-03-14
===================

  * 0.0.3
  * how to use/install: link to releases

v0.0.2 / 2015-03-14
===================

  * 0.0.2
  * update package details

v0.0.1 / 2015-03-14
===================

  * 0.0.1
  * get package list first
  * install zip first
  * fix wercker syntax
  * add build steps
  * init wercker
  * export api
  * tweak readme
  * update readme
  * run right away
  * hat tip to @shanselman
  * made with <3
  * License, some comments
  * update readme
  * 
  * cleanup output a bit
  * tweak ws
  * pass options to fetcher
  * push to ns
  * ignore things
  * tweak console
  * do stuff
  * Merge branch 'master' of gist.github.com:/45252f6e87abd0ee7a8f
  * 
  * init


tests/fetch.test.js
const should = require('should');
const accountName = process.env["DEXCOM_ACCOUNT_NAME"]
const password = process.env["DEXCOM_PASSWORD"]

describe("fetch", () => {
	let engine, sessionID;

	before(done => {
		engine = require("../index.js");
		engine.authorize(
			{
				password: password,
				applicationId: "d89443d2-327c-4a6f-89e5-496bbb0317db",
				accountName: accountName
			},
			(error, response, body) => {
				sessionID = body;
				done(error);
			}
		)
	});

	it("should return a list of recent entries when called with valid session id", done => {
		const maxCount = 5;
		engine.fetch(
			{
				sessionID,
				maxCount
			},
			(error, response, body) => {
				response.statusCode.should.equal(200);
				body.should
					.be.an.instanceOf(Array)
					.and.have.lengthOf(maxCount);
				done(error);
			}
		);
	});

	it("should return Bad Request response when called with invalid session id", done => {
		engine.fetch(
			{
				sessionID: "bad-session-id"
			},
			(error, response, body) => {
				response.statusCode.should.equal(400);
				done(error);
			}
		);		
	});
});

index.js
/**
 * Author: Ben West
 * https://github.com/bewest
 * Advisor: Scott Hanselman
 * http://www.hanselman.com/blog/BridgingDexcomShareCGMReceiversAndNightscout.aspx
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *
 * @description: Allows user to store their Dexcom data in their own
 * Nightscout server by facilitating the transfer of latest records
 * from Dexcom's server into theirs.
 */
var request = require('request');
var qs = require('querystring');
var crypto = require('crypto');
var meta = require('./package.json');


// Defaults
var server = "share2.dexcom.com";
var bridge = readENV('BRIDGE_SERVER')
    if (bridge && bridge.indexOf(".") > 1) {
    server = bridge;
   } 
    else if (bridge && bridge === 'EU') {
        server = "shareous2.dexcom.com";
    } 


var Defaults = {
  "applicationId":"d89443d2-327c-4a6f-89e5-496bbb0317db"
, "agent": [meta.name, meta.version].join('/')
, login: 'https://' + server + '/ShareWebServices/Services/General/LoginPublisherAccountByName'
, accept: 'application/json'
, 'content-type': 'application/json'
, LatestGlucose: 'https://' + server + '/ShareWebServices/Services/Publisher/ReadPublisherLatestGlucoseValues'
// ?sessionID=e59c836f-5aeb-4b95-afa2-39cf2769fede&minutes=1440&maxCount=1"
, nightscout_upload: '/api/v1/entries.json'
, nightscout_battery: '/api/v1/devicestatus.json'
, MIN_PASSPHRASE_LENGTH: 12
};

var DIRECTIONS = {
  NONE: 0
, DoubleUp: 1
, SingleUp: 2
, FortyFiveUp: 3
, Flat: 4
, FortyFiveDown: 5
, SingleDown: 6
, DoubleDown: 7
, 'NOT COMPUTABLE': 8
, 'RATE OUT OF RANGE': 9
};
var Trends = (function ( ) {
  var keys = Object.keys(DIRECTIONS);
  var trends = keys.sort(function (a, b) {
    return DIRECTIONS[a] - DIRECTIONS[b];
  });
  return trends;
})( );
function directionToTrend (direction) {
  var trend = 8;
  if (direction in DIRECTIONS) {
    trend = DIRECTIONS[direction];
  }
  return trend;
}
function trendToDirection (trend) {
  return Trends[trend] || Trends[0];
}

// assemble the POST body for the login endpoint
function login_payload (opts) {
  var body = {
    "password": opts.password
  , "applicationId" : opts.applicationId || Defaults.applicationId
  , "accountName": opts.accountName
  };
  return body;
}

// Login to Dexcom's server.
function authorize (opts, then) {
  var url = Defaults.login;
  var body = login_payload(opts);
  var headers = { 'User-Agent': Defaults.agent
                , 'Content-Type': Defaults['content-type']
                , 'Accept': Defaults.accept };
  var req ={ uri: url, body: body, json: true, headers: headers, method: 'POST'
           , rejectUnauthorized: false };
  // Asynchronously calls the `then` function when the request's I/O
  // is done.
  return request(req, then);
}

// Assemble query string for fetching data.
function fetch_query (opts) {
  // ?sessionID=e59c836f-5aeb-4b95-afa2-39cf2769fede&minutes=1440&maxCount=1"
  var q = {
    sessionID: opts.sessionID
  , minutes: opts.minutes || 1440
  , maxCount: opts.maxCount || 1
  };
  var url = Defaults.LatestGlucose + '?' + qs.stringify(q);
  return url;
}

// Asynchronously fetch data from Dexcom's server.
// Will fetch `minutes` and `maxCount` records.
function fetch (opts, then) {
  var url = fetch_query(opts);
  var body = "";
  var headers = { 'User-Agent': Defaults.agent
                , 'Content-Type': Defaults['content-type']
                , 'Content-Length': 0
                , 'Accept': Defaults.accept };

  var req ={ uri: url, body: body, json: true, headers: headers, method: 'POST'
           , rejectUnauthorized: false };
  return request(req, then);
}

// Authenticate and fetch data from Dexcom.
function do_everything (opts, then) {
  var login_opts = opts.login;
  var fetch_opts = opts.fetch;
  authorize(login_opts, function (err, res, body) {

    fetch_opts.sessionID = body;
    fetch(fetch_opts, function (err, res, glucose) {
      then(err, glucose);

    });
  });

}

// Map Dexcom's property values to Nightscout's.
function dex_to_entry (d) {
/*
[ { DT: '/Date(1426292016000-0700)/',
    ST: '/Date(1426295616000)/',
    Trend: 4,
    Value: 101,
    WT: '/Date(1426292039000)/' } ]
*/
  var regex = /\((.*)\)/;
  var wall = parseInt(d.WT.match(regex)[1]);
  var date = new Date(wall);
  var entry = {
    sgv: d.Value
  , date: wall
  , dateString: date.toISOString( )
  , trend: d.Trend
  , direction: trendToDirection(d.Trend)
  , device: 'share2'
  , type: 'sgv'
  };
  return entry;
}

// Record data into Nightscout.
function report_to_nightscout (opts, then) {
  var shasum = crypto.createHash('sha1');
  var hash = shasum.update(opts.API_SECRET);
  var headers = { 'api-secret': shasum.digest('hex')
                , 'Content-Type': Defaults['content-type']
                , 'Accept': Defaults.accept };
  var url = opts.endpoint + Defaults.nightscout_upload;
  var req = { uri: url, body: opts.entries, json: true, headers: headers, method: 'POST'
            , rejectUnauthorized: false };
  return request(req, then);

}

function nullify_battery_status (opts, then) {
  var shasum = crypto.createHash('sha1');
  var hash = shasum.update(opts.API_SECRET);
  var headers = { 'api-secret': shasum.digest('hex')
                , 'Content-Type': Defaults['content-type']
                , 'Accept': Defaults.accept };
  var url = opts.endpoint + Defaults.nightscout_battery;
  var body = { uploaderBattery: false };
  var req = { uri: url, body: body, json: true, headers: headers, method: 'POST'
            , rejectUnauthorized: false };
  return request(req, then);
}

function engine (opts) {

  var runs = 0;
  var failures = 0;
  function my ( ) {
    console.log('RUNNING', runs, 'failures', failures);
    if (my.sessionID) {
      var fetch_opts = Object.create(opts.fetch);
      if (runs === 0) {
        console.log('First run, fetching', opts.firstFetchCount);
        fetch_opts.maxCount = opts.firstFetchCount;
      }
      fetch_opts.sessionID = my.sessionID;
      fetch(fetch_opts, function (err, res, glucose) {
        if (res && res.statusCode < 400) {
          to_nightscout(glucose);
        } else {
          my.sessionID = null;
          refresh_token( );
        }
      });
    } else {
      failures++;
      refresh_token( );
    }
  }

  function refresh_token ( ) {
    console.log('Fetching new token');
    authorize(opts.login, function (err, res, body) {
      if (!err && body && res && res.statusCode == 200) {
        my.sessionID = body;
        failures = 0;
        my( );
      } else {
        failures++;
        var responseStatus = res ? res.statusCode : "response not found";
        console.log("Error refreshing token", err, responseStatus, body);
        if (failures >= opts.maxFailures) {
          throw "Too many login failures, check DEXCOM_ACCOUNT_NAME and DEXCOM_PASSWORD";
        }
      }
    });
  }

  function to_nightscout (glucose) {
    var ns_config = Object.create(opts.nightscout);
    if (glucose) {
      runs++;
      // Translate to Nightscout data.
      var entries = glucose.map(dex_to_entry);
      console.log('Entries', entries);
      if (opts && opts.callback && opts.callback.call) {
        opts.callback(null, entries);
      }
      if (ns_config.endpoint) {
        if (runs === 0) {
          nullify_battery_status(ns_config, function (err, resp) {
            if (err) {
              console.warn('Problem reporting battery', arguments);
            } else {
              console.log('Battery status hidden');
            }
          });
        }
        ns_config.entries = entries;
        // Send data to Nightscout.
        report_to_nightscout(ns_config, function (err, response, body) {
          console.log("Nightscout upload", 'error', err, 'status', response.statusCode, body);

        });
      }
    }
  }

  my( );
  return my;
}

// Provide public, testable API
engine.fetch = fetch;
engine.authorize = authorize;
engine.authorize_fetch = do_everything;
module.exports = engine;

function readENV(varName, defaultValue) {
    //for some reason Azure uses this prefix, maybe there is a good reason
    var value = process.env['CUSTOMCONNSTR_' + varName]
        || process.env['CUSTOMCONNSTR_' + varName.toLowerCase()]
        || process.env[varName]
        || process.env[varName.toLowerCase()];

    return value || defaultValue;
}

// If run from commandline, run the whole program.
if (!module.parent) {
  if (readENV('API_SECRET').length < Defaults.MIN_PASSPHRASE_LENGTH) {
    var msg = [ "API_SECRET environment variable should be at least"
              , Defaults.MIN_PASSPHRASE_LENGTH, "characters" ];
    var err = new Error(msg.join(' '));
    throw err;
    process.exit(1);
  }
  if (readENV('DEXCOM_ACCOUNT_NAME', '@').match(/\@/)) {
    var msg = [ "environment variable"
              , "DEXCOM_ACCOUNT_NAME should be"
              , "Dexcom Share user name, not an email address"];
    var err = new Error(msg.join(' '));
    throw err;
    process.exit(1);
  }
  var args = process.argv.slice(2);
  var config = {
    accountName: readENV('DEXCOM_ACCOUNT_NAME')
  , password: readENV('DEXCOM_PASSWORD')
  };
  var ns_config = {
    API_SECRET: readENV('API_SECRET')
  , endpoint: readENV('NS', 'https://' + readENV('WEBSITE_HOSTNAME'))
  };
  var interval = readENV('SHARE_INTERVAL', 60000 * 2.5);
  interval = Math.max(60000, interval);
  var fetch_config = { maxCount: readENV('maxCount', 1)
    , minutes: readENV('minutes', 1440)
  };
  var meta = {
    login: config
  , fetch: fetch_config
  , nightscout: ns_config
  , maxFailures: readENV('maxFailures', 3)
  , firstFetchCount: readENV('firstFetchCount', 3)
  };
  switch (args[0]) {
    case 'login':
      authorize(config, console.log.bind(console, 'login'));
      break;
    case 'fetch':
      config = { sessionID: args[1] };
      fetch(config, console.log.bind(console, 'fetched'));
      break;
    case 'testdaemon':
      setInterval(engine(meta), 2500);
      break;
    case 'run':
      // Authorize and fetch from Dexcom.
      do_everything(meta, function (err, glucose) {
        console.log('From Dexcom', err, glucose);
        if (glucose) {
          // Translate to Nightscout data.
          var entries = glucose.map(dex_to_entry);
          console.log('Entries', entries);
          if (ns_config.endpoint) {
            ns_config.entries = entries;
            // Send data to Nightscout.
            report_to_nightscout(ns_config, function (err, response, body) {
              console.log("Nightscout upload", 'error', err, 'status', response.statusCode, body);

            });
          }
        }
      });
      break;
    default:
      setInterval(engine(meta), interval);
      break;
      break;
  }
}

